### DISCUSSION TITLE
How to improve the convergence?

### DISCUSSION THREAD

**Discussion by zhang-rite** [idaholab/moose#17091] [Q&A Modules: General] [answered]
Hi MOOSE developers and users

Now I test a 2D RZ hydro-mechanical model.
However, the problem cannot get converged. Could you help me point out the problem. Thank you.
Here is the github [link](https://github.com/zhang-rite/MOOSE_learning/blob/main/Initialization_HM_test.i)

**Comment by WilkAndy** [comment]
Have you read these pages:

https://mooseframework.inl.gov/modules/porous_flow/solvers.html
https://mooseframework.inl.gov/modules/porous_flow/convergence.html
https://mooseframework.inl.gov/modules/porous_flow/nonlinear_convergence_problems.html

?

a

**Reply by zhang-rite** [reply]
Andy, thank you again. 
Now I have read the recommended pages and modified the input file slightly.
Now the problem can be converged to a value (with residual for pressure =43.8) but cannot be reduced further, no matter how I change the options.
(One error previously I made was misuse of "z" instead of "y" for rz coordinate).
Thanks!

**Reply by zhang-rite** [reply]
This is a steady run!

**Reply by zhang-rite** [reply]
The "disp_" looks strange. The pattern of pressure and strain seem good.
I cannot find any improper setting of BCs in the model.
```
[BCs]
  [top]
    type = FunctionDirichletBC
    variable = pwater
    boundary = 'top'
    # use_displaced_mesh = false
    function = '101325-y*9810'
  []    
     
  [xmax_drained]
    type = FunctionDirichletBC
    variable = pwater
    boundary = 'right'
    function = '101325-y*9810'  
    # use_displaced_mesh = false 
  []  

  [./fixed_outer_r]
    type = PresetBC
    variable = disp_r
    value = 0
    boundary = right
    # use_displaced_mesh = false
  [../]

  [./fixed_outer_l]
    type = PresetBC
    variable = disp_r
    value = 0
    boundary = 'left' 
    # use_displaced_mesh = false
  [../]

  [./fixed_bottom_z]
    type = PresetBC
    variable = disp_z
    value = 0
    boundary = bottom
    # use_displaced_mesh = false
  [../]  

  [./top_z]
    type = Pressure
    variable = disp_z
    component = 1
    function = '-2386.0*9.81*y'
    use_displaced_mesh = false
    boundary = 'top'
  [../]  
[]

```
```
[Preconditioning]
  [./test]
    type = SMP
    full = true
    petsc_options_iname = '-ksp_type -pc_type -sub_pc_type -pc_factor_shift_type  -snes_atol -snes_rtol'
    petsc_options_value = 'gmres      lu       mumps        NONZERO                4.4e1     1E-10'
    petsc_options = '-snes_converged_reason -ksp_diagonal_scale -ksp_diagonal_scale_fix -ksp_gmres_modifiedgramschmidt -snes_linesearch_monitor'

  [../]
[]
[Executioner]
  type = Steady
  solve_type = Newton
  l_max_its  = 200
  nl_max_its = 500
  automatic_scaling = True
  compute_scaling_once = False
[]

```
![image](https://user-images.githubusercontent.com/79129885/108816729-0de02800-75fa-11eb-9d86-0389b2b9ea16.png)

**Comment by WilkAndy** [comment, marked as answer]
@zhang-rite , i had a look at your input file.  It obviously uses an old version of MOOSE (it does not even run with modern MOOSE) which is not recommended.  Anyway, modernising the input file isn't hard and doesn't make any difference to the result in this case.  

The overall point is: **you have converged!** The solution is found, to the precision available on your computer.

You can probably squeeze better precision from the problem by

- altering the permeability (the steady-state doesn't depend on perm in this case) or elasticity tensor , since these things are so small/big that they result in bad precision loss
- paying particular attention to `scaling` (eg, perhaps set `scaling = 1E-6` for the `disp_r` and `disp_z` terms, and don't use `automatic_scaling`)
- perhaps some flags in the Kernels (eg `full_upwind_threshold` and `fallback_scheme`)

but the result won't be much different.  If you're unhappy with the result then you'll have to muck around with the mesh (large elements obviously have different contributions to the residual than small elements), or maybe do a subsequent transient run with super long time steps (won't make much difference but might make you feel better), or similar.

**Reply by WilkAndy** [reply]
I just want to add: remember your volume is huge (~1E12) so the minimum residual as listed at https://mooseframework.inl.gov/modules/porous_flow/convergence.html will be quite large

**Reply by zhang-rite** [reply]
Andy, many thanks! Following the suggestions, the residual is slightly reduced. 
Now the transient run (without fluid injection) following the stead-state run still has variable changes, but just slightly. Thus it is ok for continued fluid injection.
